docs: add overview for third parties (#25832)
nuxt/nuxt: docs/1.getting-started/5.third-parties.md

- useScript: Load any third-party script with server-side rendering support and a proxied API.
- useStyle: Load any third-party stylesheet using various asset strategies.
- Third Party Wrappers: Wrapper components and composables that make it easier to include different popular third-parties efficiently.

## How third-parties can impact user experience

Third-parties are external resources included, but not directly controlled, by a site owner to add new functionality to a website. Popular examples of third-parties include analytics, video embeds, maps, and social media integrations. Typically, third-party providers offer code snippets that can be added to the `head` or `body` section of the document.

Adding a single third-party resource to your Nuxt application might not have any noticeable impact on performance, but they can quickly begin to affect user experience if you’re not careful. Many third-parties, especially scripts, can take a relatively long time to download and execute, which can delay user interactivity and block page rendering.

> Data from the Chrome User Experience Report shows that Nuxt sites that load more third-party resources have lower [Interaction to Next Paint](https://web.dev/articles/inp) (INP) and [Largest Contentful Paint](https://web.dev/articles/lcp) (LCP) pass rates.

Reviewer: Can we be more specific about the data (maybe in a footnote to the chart)? What month and year is it from? Was it mobile sites only? Are these origins or URLs?

Author: added meta, let me know if this is sufficient 😄

Reviewer: Looks good

![Chrome User Experience Report chart](/assets/docs/getting-started/third-parties/chart.png)

While the chart's correlation does not automatically indicate causation, lab experiments and data from the [Web Almanac](https://almanac.httparchive.org/en/2022/third-parties) provide further evidence that third-party resources significantly affect page performance.

## Optimizing third-party performance

In general, it can be difficult to determine the most optimal way to load different third-party resources. Nuxt provides a number of utilities that include various approaches to improve both the developer and user experience of loading third-parties.

> ### useScript

Reviewer: Nit: Should we consider flipping the order of the third-party wrapper and useScript sections? Generally we'd want people to default to our built-in 3P wrappers first and use `useScript` only for third-parties without a built-in. If they see this section first, they may implement e.g. GA this way without scrolling to see it has a built-in

Reviewer: Haha, didn't realize we'd already flipped the order! If we keep it as is, can we add a line here that explains this is for scripts where there isn't already a built-in wrapper and link to the bottom?

Author: had that previously too 🤣 
I can add a callout back in there though 👍

Author: added callout above this

```vue
<script setup>
useScript({
 id: 'some_id',
 src: 'https://www.example.com/script.js'
})
</script>
```

Optional triggers and asset strategies can be used to allow for more fine grained control over how, and when, the script should be loaded.

```vue
<script setup>
useScript({
 id: 'some_id',
 src: 'https://www.example.com/script.js'
 trigger: new Promise((resolve) => {
   setTimeout(() => {resolve()}, 1000)
 })
})
</script>
```

> For more details about triggers, asset strategies and other supported options, please refer to [link].

### useStyles

The `useStyles` composable enables you to load third-party stylesheets in a consistent way.

```vue
<script setup>
useStyles(['https://www.example.com/styles.css'])
</script>
```

You can use optional asset strategies to inline styles directly in the `head` section of your `document`.

```vue
<script setup>
useStyles(
  ['https://www.example.com/styles.css'],
  { assetStrategy: 'inline' }
)
</script>
```

> For more details, please refer to [link].

### Third Party Wrappers

> A collection of utilities that are already configured to load and initialize popular third-parties in the most performant way . These include:

Reviewer: ```suggestion
A collection of utilities that are already configured to load and initialize popular third-parties in the most performant way. These include:
```

> Composables to fetch scripts such as `useGoogleAnalytics` and `useGoogleTagManager`
> Components to load embeds such as `<GoogleMaps>` and `<YoutubeEmbed>`

Reviewer: It doesn't look like the bullet points came through?

```suggestion
* Composables to fetch scripts such as `useGoogleAnalytics` and `useGoogleTagManager`
* Components to load embeds such as `<GoogleMaps>` and `<YoutubeEmbed>`
```

> More improvements will be made to these existing utilities in the near future to further minimize their impact on user experience.

Reviewer: Nit: I don't think it's necessary to predict future features in docs
```suggestion
```

- Third Party Wrappers: Wrapper components and composables that make it easier to include different popular third-parties efficiently. 

## How third-parties can impact user experience

Third-parties are external resources included, but not directly controlled, by a site owner to add new functionality to a website. Popular examples of third-parties include analytics, video embeds, maps, and social media integrations. Typically, third-party providers offer code snippets that can be added to the `head` or `body` section of the document. 

Adding a single third-party resource to your Nuxt application might not have any noticeable impact on performance, but they can quickly begin to affect user experience if you’re not careful. Many third-parties, especially scripts, can take a relatively long time to download and execute, which can delay user interactivity and block page rendering.

Data from the Chrome User Experience Report shows that Nuxt sites that load more third-party resources have lower [Interaction to Next Paint](https://web.dev/articles/inp) (INP) and [Largest Contentful Paint](https://web.dev/articles/lcp) (LCP) pass rates.

> ![Chrome User Experience Report chart](/assets/docs/getting-started/third-parties/chart.png)

Author: I followed other doc examples that included an image. I don't know where we actually add the asset and need some guidance here...

Author: just saw response 😄 
will open pr for nuxt.com in a few..

---
title: Third Parties
description: Learn how to optimize the performance of third-party resources using built-in composables and components.
navigation.icon: i-ph-file-search-duotone
---

Nuxt provides a number of composables and built-in libraries that make it easier, and faster, to load third-party resources in your application.

> - useScript: Load any third-party script with server-side rendering support and a proxied API.
> - useStyle: Load any third-party stylesheet using various asset strategies.

Reviewer 2: ```suggestion
- `useScript`: Load any third-party script with server-side rendering support and a proxied API.
- `useStyle`: Load any third-party stylesheet using various asset strategies.
```

> - `useStyle`: Load any third-party stylesheet using various asset strategies.

Reviewer 3: I think we can keep these composables as internal in relation to this docs page. End user does not need to know about them, this makes it easier for us to change the lower level implementation in the future if / when we need

Reviewer 3: Also the composable seems to be `useStyles` not `useStyle`

Author: I'll wait for Julien's input as well, but fixed the typo for now

> - Third Party Wrappers: Wrapper components and composables that make it easier to include different popular third-parties efficiently.

Reviewer 2: ```suggestion
- Third-Party Wrappers: Wrapper components and composables that make it easier to include different popular third parties efficiently.
```

> ## How third-parties can impact user experience

Reviewer 2: ```suggestion
## How Third-Parties Can Impact User Experience
```

> Third-parties are external resources included, but not directly controlled, by a site owner to add new functionality to a website. Popular examples of third-parties include analytics, video embeds, maps, and social media integrations. Typically, third-party providers offer code snippets that can be added to the `head` or `body` section of the document.

Reviewer 2: ```suggestion
Third parties are external resources included, but not directly controlled, by a site owner to add new functionality to a website. Popular examples of third parties include analytics, video embeds, maps, and social media integrations. Typically, third-party providers offer code snippets that can be added to the `head` or `body` section of the document.
```

> Adding a single third-party resource to your Nuxt application might not have any noticeable impact on performance, but they can quickly begin to affect user experience if you’re not careful. Many third-parties, especially scripts, can take a relatively long time to download and execute, which can delay user interactivity and block page rendering.

Reviewer 2: ```suggestion
Adding a single third-party resource to your Nuxt application might not have any noticeable impact on performance, but it can quickly begin to affect user experience if you’re not careful. Many third parties, especially scripts, can take a relatively long time to download and execute, which can delay user interactivity and block page rendering.
```

Data from the Chrome User Experience Report shows that Nuxt sites that load more third-party resources have lower [Interaction to Next Paint](https://web.dev/articles/inp) (INP) and [Largest Contentful Paint](https://web.dev/articles/lcp) (LCP) pass rates.

![Chrome User Experience Report chart](/assets/docs/getting-started/third-parties/chart.png)
<sub><sub>
source: Chrome User Experience Report,
date: October 2023,
devices: phone
</sub></sub>

While the chart's correlation does not automatically indicate causation, lab experiments and data from the [Web Almanac](https://almanac.httparchive.org/en/2022/third-parties) provide further evidence that third-party resources significantly affect page performance.

> ## Optimizing third-party performance

Reviewer 2: ```suggestion
## Optimizing Third Party Performance
```

> In general, it can be difficult to determine the most optimal way to load different third-party resources. Nuxt provides a number of utilities that include various approaches to improve both the developer and user experience of loading third-parties.

Reviewer 2: ```suggestion
In general, it can be difficult to determine the most optimal way to load different third-party resources. Nuxt provides a number of utilities that include various approaches to improve both the developer and user experience of loading third parties.
```

### useScript

The `useScript` composable enables you to load third-party scripts with SSR support and a proxied API. It works out of the box, requiring just an ID and a URL that points to the script resource.

```vue

> <script setup>
> useScript({
>  id: 'some_id',
>  src: 'https://www.example.com/script.js'
> })
> </script>

Reviewer 2: ```suggestion
<script setup lang="ts">
useScript({
  id: 'some_id',
  src: 'https://www.example.com/script.js'
})
</script>
```

```

> Optional triggers and asset strategies can be used to allow for more fine grained control over how, and when, the script should be loaded.

Reviewer 2: ```suggestion
Optional triggers and asset strategies can be used to allow for more fine-grained control over how, and when, the script should be loaded.
```

```vue

> <script setup>
> useScript({
>  id: 'some_id',
>  src: 'https://www.example.com/script.js'
>  trigger: new Promise((resolve) => {
>    setTimeout(() => {resolve()}, 1000)
>  })
> })
> </script>

Reviewer 2: ```suggestion
<script setup lang="ts">
useScript({
  id: 'some_id',
  src: 'https://www.example.com/script.js',
  trigger: new Promise((resolve) => {
    setTimeout(() => { resolve() }, 1000)
  })
})
</script>
```

```

> For more details about triggers, asset strategies and other supported options, please refer to [link].

### useStyles

> The `useStyles` composable enables you to load third-party stylesheets in a consistent way.

Reviewer 2: ```suggestion
The `useStyles` composable enables you to load third-party stylesheets consistently.
```

```vue
<script setup>
useStyles(['https://www.example.com/styles.css'])
</script>
```

You can use optional asset strategies to inline styles directly in the `head` section of your `document`.

```vue

> <script setup>

Reviewer 2: ```suggestion
<script setup lang="ts">
```

```vue
<script setup lang="ts">

> useStyles(

Reviewer 3: The issue with promoting these lower level composables, especially in this example, is that the lower level implementation at some point should be parsing and resolving this stylesheet at _build-time_ instead of the current runtime implementation

)
</script>
```

> For more details, please refer to [link].

### Third Party Wrappers

> A collection of utilities that are already configured to load and initialize popular third-parties in the most performant way. These include:

Reviewer 2: ```suggestion
A collection of utilities that are already configured to load and initialize popular third parties in the most performant way. These include:
```

---
title: Third Parties
description: Learn how to optimize the performance of third-party resources using built-in composables and components.
navigation.icon: i-ph-users-duotone
---

> Nuxt provides a number of composables and built-in libraries that make it easier, and faster, to load third-party resources in your application.

Reviewer: Super nit: em dash is more grammatically correct

```suggestion
Nuxt provides a number of composables and built-in libraries that make it easier -- and faster -- to load third-party resources in your application.
```

- `useScript`: Load any third-party script with server-side rendering support and a proxied API.
- `useStyle`: Load any third-party stylesheet using various asset strategies.
- Third-Party Wrappers: Wrapper components and composables that make it easier to include different popular third parties efficiently.

> ## How Third-Parties Can Impact User Experience

Reviewer 2: ```suggestion
## How Third Parties Can Impact User Experience
```

Third parties are external resources included, but not directly controlled, by a site owner to add new functionality to a website. Popular examples of third parties include analytics, video embeds, maps, and social media integrations. Typically, third-party providers offer code snippets that can be added to the `head` or `body` section of the document.

Adding a single third-party resource to your Nuxt application might not have any noticeable impact on performance, but it can quickly begin to affect user experience if you’re not careful. Many third parties, especially scripts, can take a relatively long time to download and execute, which can delay user interactivity and block page rendering.

Data from the Chrome User Experience Report shows that Nuxt sites that load more third-party resources have lower [Interaction to Next Paint](https://web.dev/articles/inp) (INP) and [Largest Contentful Paint](https://web.dev/articles/lcp) (LCP) pass rates.

![Chrome User Experience Report chart](/assets/docs/getting-started/third-parties/chart.png)

> <sub><sub>

Reviewer 4: There's a double sub here 🤔

source: Chrome User Experience Report,
date: October 2023,
devices: phone
</sub>

While the chart's correlation does not automatically indicate causation, lab experiments and data from the [Web Almanac](https://almanac.httparchive.org/en/2022/third-parties) provide further evidence that third-party resources significantly affect page performance.

## Optimizing Third Party Performance

> In general, it can be difficult to determine the most optimal way to load different third-party resources. Nuxt provides a number of utilities that include various approaches to improve both the developer and user experience of loading third parties.

Reviewer: ```suggestion
In general, it can be difficult to determine the most optimal way to load different third-party resources. Nuxt provides a number of utilities to improve both the developer and user experience of loading third parties.
```

::callout

> Please refer to [third-party wrappers](#third-party-wrappers) for available pre-configured solutions.

Reviewer 5: ```suggestion
Please refer to the [Third-Party Wrappers](#third-party-wrappers) section for available pre-configured solutions.
```

> Please refer to the [Third-Party Wrappers](#third-party-wrappers) section for available pre-configured solutions.

Reviewer: ```suggestion
If you're using very popular third-party libraries, please refer to the [Third-Party Wrappers](#third-party-wrappers) section for available pre-configured solutions.
```

> Please refer to [third party wrappers](#third-party-wrappers) for available pre-configured solutions.

Reviewer 2: ```suggestion
Please refer to [third-party wrappers](#third-party-wrappers) for available pre-configured solutions.
```

::

### useScript

> The `useScript` composable enables you to load third-party scripts with SSR support and a proxied API. It works out of the box, requiring just an ID and a URL that points to the script resource.

Reviewer: It's not immediately clear to me what the ID is used for. Is it something you make up to identify it? Where would you see this ID used?

Author: @huang-julien @harlan-zw - either of you want to take this one?

Reviewer: Let's just earmark to do in a follow-up then?

Reviewer 3: Not sure where the `id` has come from, you can provide a `key` when you need to load a script that has a variable loading src and you need to dedupe it i.e

```ts
useScript({
  key: 'dedupe-me',
  src: `https://example.com/script.js?time=${time}`
})
```

Otherwise it uses the script `src` to dedupe automatically. Providing an `id` will just add an `id` attribute.

Author: thanks - actually I'll only mention src, since key is optional. Might be less confusing.

```vue
<script setup lang="ts">
useScript({
  id: 'some_id',
  src: 'https://www.example.com/script.js'
})
</script>
```

> Optional triggers and asset strategies can be used to allow for more fine-grained control over how, and when, the script should be loaded.

Reviewer: I think it could be helpful to explain how the trigger relates to script loading

```suggestion
Optional triggers and asset strategies can be used to allow for more fine-grained control over how, and when, the script should be loaded. In the following example, the script will load once the promise provided in the trigger resolves (1 second later).
```

```vue
<script setup lang="ts">
useScript({
  id: 'some_id',
  src: 'https://www.example.com/script.js',
  trigger: new Promise((resolve) => {
    setTimeout(() => { resolve() }, 1000)
  })
})
</script>
```

> > For more details about triggers, asset strategies and other supported options, please refer to [link].

Author: I'll need a [link] once the api documentation for useScript is added

### useStyles

> The `useStyles` composable enables you to load third-party stylesheets consistently.

Author: @huang-julien - feel free to add a better description here 😄

Reviewer 4: maybe

```suggestion
The `useStyles` composable enables you to leverage `nuxt-assets` for third-party stylesheets.
```

?

```vue

> <script setup>

Reviewer 2: ```suggestion
<script setup lang="ts">
```

> The `useStyles` composable enables you to leverage `nuxt-assets` for third-party stylesheets.

Reviewer: Link to nuxt-assets?

Author: nuxt-assets is not in the api doc yet (@huang-julien)

useStyles(['https://www.example.com/styles.css'])
</script>
```

You can use optional asset strategies to inline styles directly in the `head` section of your `document`.

```vue
<script setup lang="ts">
useStyles(
  ['https://www.example.com/styles.css'],
  { assetStrategy: 'inline' }
)
</script>
```

> > For more details, please refer to [link].

Author: I'll need a [link] once the api documentation for useScript is added

> ### Third Party Wrappers

Reviewer 2: ```suggestion
### Third-Party Wrappers
```

> A collection of utilities that are already configured to load and initialize popular third parties in the most performant way. These include:

Reviewer: Nit: make this into a full sentence

```suggestion
Nuxt provides a collection of utilities that are already configured to load and initialize popular third parties in the most performant way. These include:
```

- Composables to fetch scripts such as `useGoogleAnalytics` and `useGoogleTagManager`
- Components to load embeds such as `<GoogleMaps>` and `<YoutubeEmbed>`

#### Google Analytics

The `useGoogleAnalytics` composable allows you to integrate Google Analytics in your Nuxt application.

```vue

> <script setup>
>   useGoogleAnalytics({
>    id: 'GA-123456789-1'
>   })
> </script>

Reviewer 2: ```suggestion
<script setup lang="ts">
useGoogleAnalytics({
  id: 'GA-123456789-1'
})
</script>
```

```

> Read more in [docs > api > composables > use-gogle-analytics](../3.api/2.composables/use-google-analytics.md)

Reviewer 2: ```suggestion
Read more in [docs > api > composables > use-gogle-analytics](../3.api/2.composables/use-google-analytics.md).
```

> Read more in [docs > api > composables > use-gogle-analytics](../3.api/2.composables/use-google-analytics.md).

Reviewer 4: ```suggestion
Read more in [docs > api > composables > use-google-analytics](../3.api/2.composables/use-google-analytics.md).
```

Reviewer 5: Looking at the other docs, looks there's already a component utility for "Read More" callouts:

```suggestion
:read-more{to="/docs/api/composables/use-google-analytics"}
```

#### Google Tag Manager

> The `useGoogleTagManager` composable makes it easier to include Google Tag Manager to your Nuxt application.

Reviewer 2: ```suggestion
The `useGoogleTagManager` composable makes it easier to include Google Tag Manager in your Nuxt application.
```

```vue

> <script setup>
>   useGoogleTagManager({
>    id: 'GTM-ABCD12'
>   })
> <script>

Reviewer 2: ```suggestion
<script setup lang="ts">
useGoogleTagManager({
  id: 'GTM-ABCD12'
})
<script>
```

```

> Read more in [docs > api > composables > use-gogle-tag-manager](../3.api/2.composables/use-google-tag-manager.md)

Reviewer 2: ```suggestion
Read more in [docs > api > composables > use-gogle-tag-manager](../3.api/2.composables/use-google-tag-manager.md).
```

> Read more in [docs > api > composables > use-gogle-tag-manager](../3.api/2.composables/use-google-tag-manager.md).

Reviewer 4: ```suggestion
Read more in [docs > api > composables > use-google-tag-manager](../3.api/2.composables/use-google-tag-manager.md).
```

Reviewer 5: See my comment above

#### Google Maps

> The `<GoogleMaps>` component initializes and displays a Google Map to your page using the Maps JavaScript API.

Reviewer 2: ```suggestion
The `<GoogleMaps>` component initializes and displays a Google Map on your page using the Maps JavaScript API.
```

```vue
​​<template>

>    <div>
>      <GoogleMaps
>        api-key="API_KEY"
>        q="Space+Needle,Seattle+WA"
>     />
>    </div>

Reviewer 2: ```suggestion
  <div>
    <GoogleMaps
      api-key="API_KEY"
      q="Space+Needle,Seattle+WA"
    />
  </div>
```

</template>
```

> Read more in [docs > api > components > google-maps](../3.api/1.components/12.google-maps.md)

Reviewer 2: ```suggestion
Read more in [docs > api > components > google-maps](../3.api/1.components/12.google-maps.md).
```

> Read more in [docs > api > components > google-maps](../3.api/1.components/12.google-maps.md).

Reviewer 5: See my comment above

#### YouTube Embed

> The `<YoutubeEmbed>` component uses `lite-youtube-embed` under the hood to load and display a YouTube embed to your page faster.

Reviewer 2: ```suggestion
The `<YoutubeEmbed>` component uses `lite-youtube-embed` under the hood to load and display a YouTube embed on your page faster.
```

```vue
​​<template>

>    <div>
>      <YoutubeEmbed
>        video-id="d_IFKP1Ofq0"
>        play-label="Play Video"
>     />
>    </div>

Reviewer 2: ```suggestion
  <div>
    <YoutubeEmbed
      video-id="d_IFKP1Ofq0"
      play-label="Play Video"
    />
  </div>
```

</template>
```

> Read more in [docs > api > components > youtube](../3.api/1.components/13.youtube-embed.md)

Reviewer 2: ```suggestion
Read more in [docs > api > components > youtube](../3.api/1.components/13.youtube-embed.md).
```

> Read more in [docs > api > components > youtube](../3.api/1.components/13.youtube-embed.md).

Reviewer 5: See my comment above
